### PR TITLE
Adds translate status feature (when connected to a Mastodon 4.x server with translation enabled)

### PR DESCRIPTION
**Pull request by danschwarz**
I've submitted a patch to the mailing list, but hopefully this is cleaner/easier to review.

Implementation as per https://github.com/mastodon/mastodon/pull/19218
and https://github.com/mastodon/mastodon/issues/19328

Screenshots of feature in action.

Screenshot 2022-12-06 192326
Screenshot 2022-12-06 192407

### CODE REVIEW

File: toot/tui/app.py [hunk 1]
```diff
@@ -303,15 +305,31 @@ def async_load_instance(self):
...
+                if "version" in instance and versiontuple(instance["version"])[0] < 4:
+                    # Mastodon versions < 4 do not have translation service
+                    self.can_translate = Option.NO
+
+            # translation service for Mastodon version 4.0.0-4.0.2 that do not advertise
+            # is indeterminate; as of now versions up to 4.0.2 cannot advertise
+            # even if they provide the service, but future versions, perhaps 4.0.3+
+            # will be able to advertise.
 
-        return self.run_in_thread(_load_instance, done_callback=_done)
```
Comment (ihabunek) [inline toot/tui/app.py:314]: You deleted this line so we're never running above code.

**Review by ihabunek — CHANGES_REQUESTED** [review]
Mailing list is fine, as are pulls on Github. The patch you sent to the mailing list does not apply, so check it out.

Thanks, this is a good patch, I have some comments before we merge.

File: toot/tui/timeline.py [hunk 2]
```diff
@@ -157,7 +159,12 @@ def keypress(self, size, key):
             self._emit("links", status)
             return
 
-        if key in ("t", "T"):
+        if key in ("n", "N"):
+            if self.can_translate != Option.NO:
+                self._emit("translate", status)
+            return
+
+        if key in ("r", "R"):
```
[the remark this answers is not in the document]
Comment (danschwarz) [reply toot/tui/timeline.py]: Ouch. Not sure how that happened. Fixed.

File: toot/tui/app.py [hunk 3]
```diff
@@ -106,6 +106,7 @@ def __init__(self, app, user):
         self.timeline = None
         self.overlay = None
         self.exception = None
+        self.can_translate = Option.UNKNOWN
```
[the remark this answers is not in the document]
Comment (danschwarz) [reply toot/tui/app.py]: Done

File: toot/tui/app.py [hunk 4]
```diff
@@ -303,15 +305,31 @@ def async_load_instance(self):
...
         """
         def _load_instance():
             return api.get_instance(self.app.instance)
 
         def _done(instance):
             if "max_toot_chars" in instance:
                 self.max_toot_chars = instance["max_toot_chars"]
+            if "translation" in instance:
+                # instance is advertising translation service
+                self.can_translate = Option.YES
```
[the remark this answers is not in the document]
Comment (danschwarz) [reply toot/tui/app.py]: Fixed

File: toot/tui/app.py (hunk 1 above)
Comment (danschwarz) [reply toot/tui/app.py:314]: Added it back. Thanks.

**Comment by danschwarz** [comment]
I've pushed fixes for all of the above issues, hopefully all is good now. Enhancements for future consideration:

* Use a library to translate language codes ('en', 'de') to language names when displaying "translated from..." text
* Make translations toggleable

**Comment by ihabunek** [comment]
Thanks! I merged this manually and also enabled toggling translation by pressing n repeatedly.

**Comment by ihabunek** [comment]
>     * Use a library to translate language codes ('en', 'de') to language names when displaying "translated from..." text

**Comment by ihabunek** [comment]
>     * Use a library to translate language codes ('en', 'de') to language names when displaying "translated from..." text

We don't need a library, when we can use language codes from wikipedia: https://github.com/ihabunek/toot/commit/0ab0db048cd12a38da48eee8d406e8af883e8713

**Pull request closed without merging**